File: Item.py
from mc_base import *
from typing import Self

class Item:
    item_cache = {}
    def __init__(self, item: str|dict|LLSE_Item|NbtCompound|Self, count: int = None) -> None:
        self.item_dict: dict = {}
        
        if isinstance(item, Item):
            self.item_dict = item.item_dict.copy()
            if count is not None:
                self.count = count
            return
        
        if isinstance(item, (dict, NbtCompound)):
            self.item_dict = item.copy() if isinstance(item, dict) else item.toObject()
            self.item_dict.pop('Slot', None)
            if count is not None:
                self.item_dict['Count'] = count
        elif isinstance(item, LLSE_Item):
            count = count if count is not None else 1
            self.item_dict = item.getNbt().toObject()
            self.item_dict.pop('Slot', None)
        elif isinstance(item, str):
            count = count if count is not None else 1
            if item in self.item_cache:
                self.item_dict = self.item_cache[item].copy()
            else:
                llse_item: LLSE_Item|None = mc.newItem(item, count)
                if llse_item:
                    self.item_dict = llse_item.getNbt().toObject()
                    self.item_cache[item] = self.item_dict.copy()
                else:
                    self.item_dict = {'Count': 0, 'Name': '', 'Damage': 0, 'WasPickedUp': 0}
        
        if any(key not in self.item_dict for key in ('Count', 'Name', 'Damage', 'WasPickedUp')):
            raise ValueError(f'filed builed item for item_dict: {self.item_dict}')
        
    @property
    def count(self) -> int:
        return self.item_dict['Count']
    
    @count.setter
    def count(self, count: int) -> None:
        self.item_dict['Count'] = count
        
    @property
    def damage(self) -> int:
        return self.item_dict['Damage']
    
    @damage.setter
    def damage(self, damage: int) -> None:
        self.item_dict['Damage'] = damage
    
    @property
    def type(self) -> str:
        return self.item_dict['Name']
                  
    def match(self, llse_item: LLSE_Item|NbtCompound) -> bool:
        if isinstance(llse_item, NbtCompound):
            name_nbt = llse_item.getTag('Name') 
            if not name_nbt or name_nbt.get() != self.type:
                return False
            
            damage_nbt = llse_item.getTag('Damage') 
            if not damage_nbt or damage_nbt.get() != self.damage:
                return False
        
        elif self.type != llse_item.type or self.damage != llse_item.damage:
            return False
        
        return self.get_identifier() == self.get_llse_item_identifier(llse_item)
    
    def get_identifier(self, include_count: bool = False) -> str:
        item_dict = self.item_dict.copy()
        if not include_count:
            item_dict.pop('Count', None)
        item_dict.pop('Block', None)
        return str(item_dict).replace('{}', 'None').replace('[]', 'None')
    
    def to_llse_item(self) -> LLSE_Item|None:
        if self.type in black_item_name:
            return None
        
        item = mc.newItem(self.type, self.count) 
            
        if item:
            item_nbt = item.getNbt()
            if item_nbt:
                item_dict = self.item_dict.copy()
                item_dict.pop('Name', None)
                item_dict.pop('Count', None)
                
                item_nbt.setShort('Damage', item_dict['Damage']) 
                if 'tag' in item_dict:
                    tag_nbt = NBT.parseSNBT(str(item_dict['tag']))
                    item_nbt.setTag('tag', tag_nbt)
                    
                item.setNbt(item_nbt)
        return item
    
    @staticmethod
    def get_llse_item_identifier(llse_item: LLSE_Item|NbtCompound, include_count: bool = False) -> str:
        item_dict = llse_item.getNbt().toObject() if isinstance(llse_item, LLSE_Item) else llse_item.toObject()
        if not include_count:
            item_dict.pop('Count', None)
        item_dict.pop('Block', None)
        return str(item_dict)

    @staticmethod
    def is_item(obj) -> bool:
        if isinstance(obj, (Item, LLSE_Item)):
            return True
        
        if not isinstance(obj, dict):
            return False
        
        return all(key in obj for key in ('Count', 'Name', 'Damage', 'WasPickedUp'))

File: Item.js
class Item {
    static item_cache = {};

    constructor(item, count = null) {
        this.item_dict = {};

        if (item instanceof Item) {
            this.item_dict = Object.assign({}, item.item_dict);
        } else if (isLLSEItem(item)) {
            this.item_dict = item.getNbt().toObject();
            delete this.item_dict.Slot;
        } else if (typeof item === 'string') {
            if (item in Item.item_cache) {
                this.item_dict = Object.assign({}, Item.item_cache[item]);
            } else {
                let llse_item = mc.newItem(item, 1);
                if (llse_item) {
                    this.item_dict = llse_item.getNbt().toObject();
                    Item.item_cache[item] = Object.assign({}, this.item_dict);
                } else {
                    this.item_dict = { Count: 0, Name: '', Damage: 0, WasPickedUp: 0 };
                }
            }
        } else if (isDictOrNbt(item)) {
            this.item_dict = isNbtCompound(item) ? item.toObject() : Object.assign({}, item);
            delete this.item_dict.Slot;
        } 

        if (count !== null) {
            this.count = count;
        }

        // 验证必需字段
        const requiredKeys = ['Count', 'Name', 'Damage', 'WasPickedUp'];
        if (requiredKeys.some(key => !(key in this.item_dict))) {
            logger.log(this.item_dict, item, item instanceof Item, isDictOrNbt(item), isNbtCompound(item), isLLSEItem(item), typeof item === 'string');

            throw new Error(`Failed to build item for item_dict: ${JSON.stringify(this.item_dict)}`);
        }
    }

    get count() {
        return this.item_dict.Count;
    }

    set count(count) {
        this.item_dict.Count = count;
    }

    get damage() {
        return this.item_dict.Damage;
    }

    set damage(damage) {
        this.item_dict.Damage = damage;
    }

    get type() {
        return this.item_dict.Name;
    }

    match(llse_item) {
        if (isNbtCompound(llse_item)) {
            let name_nbt = llse_item.getTag('Name');
            if (!name_nbt || name_nbt.get() !== this.type) {
                return false;
            }
            
            let damage_nbt = llse_item.getTag('Damage');
            if (!damage_nbt || damage_nbt.get() !== this.damage) {
                return false;
            }
        } else if (this.type !== llse_item.type || this.damage !== llse_item.damage) {
            return false;
        }
        
        return this.get_identifier() === Item.get_llse_item_identifier(llse_item);
    }

    get_identifier(include_count = false) {
        let item_dict = Object.assign({}, this.item_dict);
        if (!include_count) {
            delete item_dict.Count;
        }
        delete item_dict.Block;
        return JSON.stringify(item_dict).replace('{}', 'null').replace('[]', 'null');
    }

    to_llse_item() {
        if (black_item_name && black_item_name.has(this.type)) {
            return null;
        }
        
        let item = mc.newItem(this.type, this.count);
        if (item) {
            let item_nbt = item.getNbt();
            if (item_nbt) {
                let item_dict = Object.assign({}, this.item_dict);
                delete item_dict.Name;
                delete item_dict.Count;
                
                item_nbt.setShort('Damage', item_dict.Damage);
                if ('tag' in item_dict) {
                    let tag_nbt = NBT.parseSNBT(JSON.stringify(item_dict.tag));
                    item_nbt.setTag('tag', tag_nbt);
                }
                item.setNbt(item_nbt);
            }
        }
        return item;
    }

    static get_llse_item_identifier(llse_item, include_count = false) {
        let item_dict = llse_item.getNbt ? llse_item.getNbt().toObject() : llse_item.toObject();
        if (!include_count) {
            delete item_dict.Count;
        }
        delete item_dict.Block;
        delete item_dict.Slot;
        return JSON.stringify(item_dict).replace('{}', 'null').replace('[]', 'null');
    }

    static is_item(obj) {
        if (obj instanceof Item || isLLSEItem(obj)) {
            return true;
        }
        
        if (!obj || typeof obj !== 'object') {
            return false;
        }
        
        const requiredKeys = ['Count', 'Name', 'Damage', 'WasPickedUp'];
        return requiredKeys.every(key => key in obj);
    }
}

// 辅助函数
function isNbtCompound(obj) {
    return obj instanceof NbtCompound;
}

function isLLSEItem(obj) {
    return obj instanceof LLSE_Item;  // 如果 LLSE_Item 也全局可用
}

function isDictOrNbt(obj) {
    if (!obj) { return false;}
    
    if (obj instanceof NbtCompound) {return true;}
    
    return obj !== null &&
           typeof obj === 'object' &&
           !Array.isArray(obj) &&
           !(obj instanceof Date) &&
           !(obj instanceof RegExp) &&
           obj.constructor === Object;
}